crates/amaru-sim/src/bin/amaru-sim/sync.rs

```diff
@@ -0,0 +1,194 @@
+use amaru_kernel::Point;
+use amaru_ouroboros::protocol::{peer::Peer, PullEvent};
+use gasket::framework::*;
+use maelstrom::protocol::Message;
+use serde::{Deserialize, Serialize};
+use std::io;
+use tracing::{error, trace_span, Span};
+
+use crate::bytes::Bytes;
+
+pub type DownstreamPort = gasket::messaging::OutputPort<PullEvent>;
+
+pub enum WorkUnit {
+    Send(PullEvent),
+}
+
+#[derive(Stage)]
+#[stage(name = "pull", unit = "WorkUnit", worker = "Worker")]
+pub struct Stage {
+    pub downstream: DownstreamPort,
+}
+
+impl Stage {
+    pub(crate) fn new(_tip: &Point) -> Self {
+        Self {
+            downstream: DownstreamPort::default(),
+        }
+    }
+}
+
+pub struct Worker {}
+
+#[async_trait::async_trait(?Send)]
+impl gasket::framework::Worker<Stage> for Worker {
+    async fn bootstrap(_stage: &Stage) -> Result<Self, WorkerError> {
+        let worker = Self {};
+
+        Ok(worker)
+    }
+
+    async fn schedule(
+        &mut self,
+        _stage: &mut Stage,
+    ) -> Result<WorkSchedule<WorkUnit>, WorkerError> {
+        // read one line of stdin which should be a JSON-formatted message from
+        // some peer to our peer
+        let mut input = String::new();
+        let span = trace_span!("pull-worker");
+
+        io::stdin()
+            .read_line(&mut input)
+            .map_err(|_| WorkerError::Recv)?;
+
+        match serde_json::from_str::<Message>(input.as_str()) {
+            Ok(v) => {
+                let msg = mk_message(v, span)?;
+                Ok(WorkSchedule::Unit(WorkUnit::Send(msg)))
+            }
+            Err(err) => {
+                error!("failed to deserialize input {}", err);
+                Err(WorkerError::Recv)
+            }
+        }
+    }
+
+    async fn execute(&mut self, unit: &WorkUnit, stage: &mut Stage) -> Result<(), WorkerError> {
+        match unit {
+            WorkUnit::Send(event) => stage.downstream.send(event.clone().into()).await.or_panic(),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Deserialize, Serialize)]
+pub struct Fwd {
+    slot: u64,
+    hash: Bytes,
+    header: Bytes,
```

Comment:
It feels somewhat redundant (and potentially error-prone) to have both hash and header ? Why not just the header?

Reply:
The `Fwd` message is used to represent arbitrary input coming from the network. I don't think there's anything in the protocol ensuring the `Point` and the `Header` that are provided in the `RollForward` messages are consistent. Keeping both the `hash` and `header` will allow us to simulate the case of faulty messages.

Reply:
Indeed, it's not strictly necessary as the underlying network message is just a `Header`. It just happens I have started using the `PullEvent::RollForward` structure which contains the `Point` so it needs the hash. I could refactor that structure and see where it leads me...